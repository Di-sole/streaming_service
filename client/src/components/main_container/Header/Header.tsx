import Link from 'next/link';
import Image from 'next/image';
import appleIcon from '../../../../public/icons/logo.svg';
import searchIcon from '../../../../public/icons/search.svg';
import bellIcon from '../../../../public/icons/bell.svg';
import userIcon from '../../../../public/icons/user.svg';
import LinksList from '../LinksList/LinksList';
import styles from './Header.module.scss';
import DropMenu from '../DropMenu/DropMenu';
import { useEffect, useState } from 'react';
import { useRouter } from 'next/router';
import TextLinkUI from '@/components/UI/links/TextLink/TextLinkUI';
import ColoredButton from '@/components/UI/buttons/ColoredButton/ColoredButton';
import { useTranslation } from 'next-i18next';
import { useAppSelector } from '@/store/hooks/redux';

interface HeaderProps {
    page: 'home' | 'other';
}

const Header = ({ page }: HeaderProps) => {
    const { t } = useTranslation('header');
    const [isShowMoviesDrop, setIsShowMoviesDrop] = useState<boolean>(false);
    const [isMouseOver, setIsMouseOver] = useState<boolean>(false);

    const { asPath, locale } = useRouter();

    const currentLocale = locale === 'ru' ? 'en' : 'ru';

    const clearShowHandler = () => {
        setIsShowMoviesDrop(false);
    };

    const mouseOverMovieHandler = () => {
        clearShowHandler();
        setIsShowMoviesDrop(true);
    };

    const clearMouseHandlers = () => {
        clearShowHandler();
    };

    useEffect(() => {
        if (isShowMoviesDrop) {
            setIsMouseOver(true);
        } else {
            setIsMouseOver(false);
        }
    }, [isShowMoviesDrop]);

    const { genres } = useAppSelector((state) => state.staticData);

    const genersLinks = () => {
        const array: {
            text: string;
            href: string;
        }[] = [];
        genres.forEach((genre) => {
            array.push({ text: locale == 'ru' ? genre.name : genre.enName, href: `/collections/${genre.enName}` });
        });
        return array;
    };

    return (
        <div onMouseLeave={clearMouseHandlers} className="container">
            <div
                className={[
                    styles.container,
                    styles[`container_${page}`],
                    isMouseOver ? styles.container_mouseOver : '',
                ].join(' ')}
            >
                <Link className={styles.container__logo} href={'/'}>
                    <Image src={appleIcon} alt="ivi"></Image>
                </Link>

                <div className={styles.container__menuList}>
                    <LinksList
                        direction="row"
                        links={[
                            {
                                href: '/',
                                text: t('myIvi'),
                                option: 'dim',
                                onMouseOver: clearShowHandler,
                            },
                            {
                                href: 'https://www.ivi.ru/new',
                                text: t('latest'),
                                option: 'dim',
                                onMouseOver: clearShowHandler,
                                target: '_blank',
                            },
                            {
                                href: '/movies',
                                text: t('movies'),
                                option: 'dim',
                                onMouseOver: mouseOverMovieHandler,
                                className: [styles.container__menuLink, styles.container__menuLink_movie].join(' '),
                            },
                            {
                                href: 'https://www.ivi.ru/series',
                                text: t('tvShows'),
                                option: 'dim',
                                onMouseOver: clearShowHandler,
                                target: '_blank',
                            },
                            {
                                href: 'https://www.ivi.ru/tvplus',
                                text: t('broadcast'),
                                option: 'dim',
                                onMouseOver: clearShowHandler,
                                target: '_blank',
                            },
                        ]}
                    />
                </div>

                <div className={styles.container__rightSide}>
                    <ColoredButton color="red" size="small" className={styles.container__button}>
                        {t('freePeriod')}
                    </ColoredButton>

                    <TextLinkUI href={`${asPath}?ivi_search`} option="bright" className={styles.container__search}>
                        <Image className={styles.container__icon} src={searchIcon} alt="" />
                        <p className={styles.container__text}>{t('search')}</p>
                    </TextLinkUI>

                    <Link href={'/'} className={[styles.container__link, styles.container__link_bell].join(' ')}>
                        <Image className={styles.container__icon} src={bellIcon} alt="Уведомления" />
                    </Link>

                    <Link
                        href={asPath}
                        locale={currentLocale}
                        className={[styles.container__link, styles.container__link_locale].join(' ')}
                    >
                        {currentLocale}
                    </Link>

                    <Link
                        href={`${asPath}?sign-in`}
                        className={[styles.container__link, styles.container__link_user].join(' ')}
                    >
                        <Image
                            className={[styles.container__icon, styles.container__icon_user].join(' ')}
                            src={userIcon}
                            alt="Пользователь"
                        />
                    </Link>
                </div>

                <DropMenu
                    className={[styles.container__dropMenu, isMouseOver ? styles.container__dropMenu_show : ''].join(
                        ' ',
                    )}
                    content={
                        isShowMoviesDrop
                            ? [
                                  {
                                      title: t('dropmenu.genres'),
                                      links: genersLinks(),
                                  },
                                  {
                                      title: t('dropmenu.countries'),
                                      links: [
                                          {
                                              text: locale === 'ru' ? 'Русские' : 'Russian',
                                              href: '/collections/Russia',
                                          },
                                          {
                                              text: locale === 'ru' ? 'Зарубежные' : 'Foreign',
                                              href: '/collections/foreign',
                                          },
                                          {
                                              text: locale === 'ru' ? 'Советское кино' : 'USSR movie',
                                              href: '/collections/USSR',
                                          },
                                      ],
                                  },
                                  {
                                      title: t('dropmenu.years'),
                                      links: [
                                          {
                                              text: locale === 'ru' ? 'Фильмы 2023 года' : 'Movies 2023',
                                              href: '/collections/2023',
                                          },
                                          {
                                              text: locale === 'ru' ? 'Фильмы 2022 года' : 'Movies 2022',
                                              href: '/collections/2022',
                                          },
                                          {
                                              text: locale === 'ru' ? 'Фильмы 2021 года' : 'Movies 2021',
                                              href: '/collections/2021',
                                          },
                                          {
                                              text: locale === 'ru' ? 'Фильмы 2020 года' : 'Movies 2020',
                                              href: '/collections/2020',
                                          },
                                          {
                                              text: locale === 'ru' ? 'Фильмы 2019 года' : 'Movies 2019',
                                              href: '/collections/2019',
                                          },
                                          {
                                              text: locale === 'ru' ? 'Фильмы 2018 года' : 'Movies 2018',
                                              href: '/collections/2018',
                                          },
                                      ],
                                  },
                              ]
                            : []
                    }
                />
            </div>
        </div>
    );
};

export default Header;
