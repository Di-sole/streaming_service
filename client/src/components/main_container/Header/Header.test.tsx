import { render, screen } from '@testing-library/react';
import Header from './Header';
import * as reduxHooks from '../../../store/hooks/redux';

jest.mock('../../../store/hooks/redux');
const mockedSelector = jest.spyOn(reduxHooks, 'useAppSelector');
const genres = [{ id: 1, name: 'Жанр', enName: 'Genre' }];

jest.mock('next-i18next', () => ({
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        };
    },
}));

jest.mock('next/router', () => ({
    useRouter() {
        return {
            route: '/',
            pathname: '',
            asPath: '/',
        };
    },
}));

describe('HEADER SNAPSHOTS TESTS', () => {
    test('Header should not be changed', () => {
        mockedSelector.mockReturnValue({ genres });
        render(<Header page="home" />);

        const header = screen.getByTestId('header');
        expect(header).toMatchSnapshot();
    });
});
