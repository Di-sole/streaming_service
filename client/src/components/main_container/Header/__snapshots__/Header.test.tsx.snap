// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HEADER SNAPSHOTS TESTS Header should not be changed 1`] = `
<div
  class="container"
  data-testid="header"
>
  <div
    class="container container_home "
  >
    <a
      class="container__logo"
      href="/"
    >
      <img
        alt="ivi"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
    </a>
    <div
      class="container__menuList"
    >
      <ul
        class="container container__menu"
      >
        <li
          class="container__item"
        >
          <a
            class="container container_dim"
            href="/"
          >
            myIvi
          </a>
        </li>
        <li
          class="container__item"
        >
          <a
            class="container container_dim"
            href="https://www.ivi.ru/new"
            target="_blank"
          >
            latest
          </a>
        </li>
        <li
          class="container__item"
        >
          <a
            class="container container_dim container__menuLink container__menuLink_movie"
            href="/movies"
          >
            movies
          </a>
        </li>
        <li
          class="container__item"
        >
          <a
            class="container container_dim"
            href="https://www.ivi.ru/series"
            target="_blank"
          >
            tvShows
          </a>
        </li>
        <li
          class="container__item"
        >
          <a
            class="container container_dim"
            href="https://www.ivi.ru/tvplus"
            target="_blank"
          >
            broadcast
          </a>
        </li>
      </ul>
    </div>
    <div
      class="container__rightSide"
    >
      <button
        class="button button_small button_red container__button"
        type="button"
      >
        freePeriod
      </button>
      <a
        class="container container_bright container__search"
        href="/?ivi_search"
      >
        <img
          alt=""
          class="container__icon"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <p
          class="container__text"
        >
          search
        </p>
      </a>
      <a
        class="container__link container__link_bell"
        href="/"
      >
        <img
          alt="Уведомления"
          class="container__icon"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </a>
      <a
        class="container__link container__link_locale"
        href="/"
      >
        ru
      </a>
      <a
        class="container__link container__link_user"
        href="/?sign-in"
      >
        <img
          alt="Пользователь"
          class="container__icon container__icon_user"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </a>
    </div>
    <div
      class="container container__dropMenu "
      data-testid="dropMenu"
    />
  </div>
</div>
`;
